### ISSUE TITLE
Add support for OpenWRT builds

### ISSUE THREAD

**Issue by mark-stopka** [openSUSE/open-build-service#10228]
**Is your feature request related to a problem? Please describe.**
I think the process for building OpenWRT packages is terrible,

**Describe the solution you'd like**
I would like Open Build Service which I use often to support OpenWRT packages also.

**Describe alternatives you've considered**
Building custom CD workflow for 3rd party software out of their repo, which breaks traditional distribution release workflow.

**Additional context**
I would be willing to financially contribute and I think CZ.NIC who makes Turris Omnia would also appreciate better OpenWRT system, correct @miska?

**Comment by adrianschroeter** [comment]
Most of the needed work is described here:

https://github.com/openSUSE/obs-build/blob/master/HOWTO.add_another_format

you can run local builds easily via the build script. Ask on IRC or mailing list if you need help.

**Comment by miska** [comment]
I was looking nto that, tricky part is that it is hard to make something like -devel packages out of OpenWrt build and SDK :-(

**Comment by mark-stopka** [comment]
Could we perhaps brainstorm about it for 30 minutes via a web conference?
And see if we can get somewhere of we join forces?

--
Sent from mobile
Best regards / S pozdravem,
BSc. Mark Stopka, BBA
Managing Partner @ PERLUR Group

mobile: +420 704 373 561
website: www.perlur.cloud

**Comment by miska** [comment]
Sure, I might have some contact information how to reach you, just us two is ok, or does somebody else want to join the discussion?

**Comment by mark-stopka** [comment]
Hi, you can reach me at mark.stopka@perlur.cloud e-mail... or we are still
friends on Facebook IMO... 😂
--
Best regards / S pozdravem,
BSc. Mark Stopka, BBA

**Comment by mark-stopka** [comment]
@adrianschroeter I think OpenWRT by default uses cross-compilation so we may avoid using Qemu for cross-platform builds and this implement in a way that is secribed as **SAFE** here, without the many risks of executing arbitrary code, @miska do I remember correctly from our last conversation about this?

**Comment by adrianschroeter** [comment]
the cross build would most likely still happen inside of any VM mode, so using KVM would make it secure.

The only problem which may appear is that you may need to track dependency for host and target independend. But you could try first without (and having always the superset of dependencies). You may need to package the openwrt packages in addition into a rpm/deb/arch in first step. So you would just need to implement the build-receipe-openwrt.

We could do native binary package format parsing and/or cross dependency resolutions in later steps.

**Comment by mark-stopka** [comment]
@pavlix, would you be able to assist?

**Comment by mark-stopka** [comment]
@vyskocilm could perhaps have the right balance of bandwidth and skill, as discussed with @miska today...

**Comment by mark-stopka** [comment]
@nijel aren't you building OpenWRT images / packages for Braiins OS+ and Braiins OS?

**Comment by nijel** [comment]
@mark-stopka I'm not involved in Braiins OS builds at all, sorry. The build system used there is at https://github.com/braiins/braiins/tree/bos-devel/braiins-os

**Comment by adrianschroeter** [comment]
@miska regarding OpenWRT build and SDK: I would suggest for the start that you put it into an opkg package manually.

We can also think about to change the preinstallimage mechanic into something what is able to install the system without the need to have it some package before (atm it is only cache mechanic).

**Comment by mark-stopka** [comment]
@nijel, thanks for the link, it looks like almost a reasonable build system for OpenWRT :-) what do you think @pavlix, @miska?

**Comment by pavlix** [comment]
As far as I know the dependency system in OpenWRT builds (SDK would work the same I guess) just calls into the build of the dependencies before building the requested package. The results are freely kept in directories that are used in subsequent build. A build service based approach might be much more reliable than the current way.

@miska Why don't we just cache the newly added files to `staging_dir` as `-devel` packages. OpenWRT already supports *host* and *target* packages, right? We could just hook into the build process and pick up everything we need to create a new type of *host* packages? Correct me if I'm wrong. I didn't check all the details.

**Comment by adrianschroeter** [comment]
the docker file makes only sense for webui development.

Please check out this instead:

https://github.com/openSUSE/obs-build/blob/master/HOWTO.add_another_format

if you add support in build (and/or pbuild) we can take care of the OBS parts easily. feel free to reach out via mailing list or IRC for further support though.

**Comment by trounticonan** [comment]
For building OpenWrt firmware (building an image to be flashed onto a router), Internet access is needed, e.g. `/scripts/feeds update -a`, and also to download the source code archives for OpenWrt packages (`make download`), but for openSUSE's Open Build Service instance, Internet access is not available during the build.

In addition, `make download` downloads most but not all necessary files.
During the compilation, you will find that some additional source code archive files will be downloaded.
In other words, if you clone the OpenWrt source code, configure feeds and build config, `make download`, and then turn off Internet access, the build will likely fail.

That being said, this feature request seems to be mainly about building individual packages, not the device firmware image, so I am not sure whether this (lack of Internet access) will be an issue.

**Comment by zoobab** [comment]
"Internet access is not available during the build."

If you use the OpenWRT SDK, you could build your package without internet access, if you have the tarball on localhost.

I am new to OBS, but I can see the tarballs are cached/stored on OBS.

**Comment by trounticonan** [comment]
As an example, if I use the SDK and build dnsmasq, `make package/feeds/base/dnsmasq/download` will only download the dnsmasq source tarball, while `make download` will also download the dependencies of dnsmasq, i.e. source tarballs of libjson-c, libmnl etc.

When building an OpenWrt package X, not only the sources of X has to be downloaded, but sources of dependencies A, B, C have to be downloaded as well for the build to succeed.

If OpenWrt support is added to OBS, how would the procedure be like?
Do I have to manually prepare the list of source tarballs?
Or will OBS automatically download them?

-----

I would be very happy to see OpenWrt firmware and/or package build support on OBS (OBS is a better platform for building packages, and is a lot faster than GitHub Actions, and GitHub Actions is more for CI I guess), but I am just wondering whether there are any technical limitations which may make it difficult to implement such support.

**Comment by ndren** [comment]
Hi!

I've been recently approved for GSoC 2023 to work on adding support for OpenWRT under the Freifunk project. I'd love to talk and learn more about how this could work! I've just set up a local build system as per https://github.com/openSUSE/open-build-service/blob/master/CONTRIBUTING.md#how-to-setup-an-obs-development-environment.

How could I set up a package for a different build recipe (for example build-recipe-arch) so I could get useful logs and keep track of what is happening at each build stage to slowly port towards build-recipe-openwrt? I could not figure out what calls build-recipe-* exactly. (In short, what is a good, simple example package I can use to verify the changes I make to the build system "work"?)

**Comment by hennevogel** [comment]
@ndren https://github.com/openSUSE/obs-build/blob/master/HOWTO.add_another_format
